File: internal/media/normalize.go
package media

import (
	"regexp"
	"strings"

	"github.com/pemistahl/lingua-go"
)

func normalizeLanguage(l lingua.Language) string {
	switch l {
	case lingua.Chinese:
		return LanguageChinese
	case lingua.Japanese:
		return LanguageJapanese
	case lingua.English:
		return LanguageEnglish
	default:
		return l.String()
	}
}

func normalizeQuality(quality string) string {
	quality = strings.ToLower(strings.ReplaceAll(quality, ".", ""))
	if strings.Contains(quality, "remux") {
		return "Remux"
	}

	switch quality {
	case "web-dl", "webdl":
		return "WEB-DL"
	case "web-rip", "webrip":
		return "WEBRip"
	case "bluray", "blu-ray":
		return "BluRay"
	case "bdrip", "bd-rip":
		return "BDRip"
	case "brrip", "br-rip":
		return "BRRip"
	case "remux":
		return "Remux"
	default:
		return quality
	}
}

func normalizeVideoCodec(codec string) string {
	codec = strings.ToUpper(codec)
	switch codec {
	case "X264", "H.264", "AVC":
		return "H264"
	case "X265", "H.265", "HEVC":
		return "H265"
	default:
		return codec
	}
}

var audioNormalizeRe = regexp.MustCompile(`\d\.\d`)

func normalizeAudioCodec(codec string) string {
	codec = strings.ToUpper(codec)
	match := audioNormalizeRe.FindStringIndex(codec)
	if match == nil {
		return parseAudioCodec(codec)
	}

	parts := []string{
		parseAudioCodec(codec[:match[0]]),
		codec[match[0]:match[1]],
	}
	left := parseAudioCodec(codec[match[1]:])
	if left != "" {
		parts = append(parts, left)
	}

	return strings.Join(parts, ".")
}

func parseAudioCodec(codec string) string {
	if codec == "" {
		return codec
	}

	parts := []string{}
	val := strings.ReplaceAll(codec, " ", ".")
	for _, p := range strings.Split(val, ".") {
		p = strings.TrimSpace(p)
		if p == "" {
			continue
		}

		switch p {
		case "TRUEHD":
			p = "TrueHD"
		case "ATMOS":
			p = "Atmos"
		case "DTSHD":
			p = "DTS-HD"
		}
		parts = append(parts, p)
	}

	return strings.Join(parts, ".")
}

func normalizeHDR(hdr string) string {
	hdr = strings.ToUpper(strings.ReplaceAll(hdr, "-", ""))
	switch {
	case strings.Contains(hdr, "DOLBY") || hdr == "DOVI":
		return "DV"
	default:
		return hdr
	}
}

File: src/domain/media/normalize.rs
use std::sync::LazyLock;

use lingua::Language;
use regex::Regex;

static AUDIO_NORMALIZE_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\d\.\d").unwrap());

/// Normalizes a language to a standard format
pub(super) fn normalize_language(l: Language) -> String {
    match l {
        Language::Chinese => super::LANGUAGE_CHINESE.to_owned(),
        Language::Japanese => super::LANGUAGE_JAPANESE.to_owned(),
        Language::English => super::LANGUAGE_ENGLISH.to_owned(),
    }
}

/// Normalizes quality string to a standard format
pub(super) fn normalize_quality(quality: &str) -> String {
    let quality = quality.to_lowercase().replace(".", "");
    if quality.contains("remux") {
        return "Remux".to_owned();
    }

    match quality.as_str() {
        "web-dl" | "webdl" => "WEB-DL".to_owned(),
        "web-rip" | "webrip" => "WEBRip".to_owned(),
        "bluray" | "blu-ray" => "BluRay".to_owned(),
        "bdrip" | "bd-rip" => "BDRip".to_owned(),
        "brrip" | "br-rip" => "BRRip".to_owned(),
        _ => quality,
    }
}

/// Normalizes video codec string to a standard format
pub(super) fn normalize_video_codec(codec: &str) -> String {
    let codec = codec.to_uppercase();
    match codec.as_str() {
        "X264" | "H.264" | "AVC" => "H264".to_owned(),
        "X265" | "H.265" | "HEVC" => "H265".to_owned(),
        _ => codec,
    }
}

/// Normalizes audio codec string to a standard format
pub(super) fn normalize_audio_codec(codec: &str) -> String {
    let codec = codec.to_uppercase();
    if let Some(m) = AUDIO_NORMALIZE_RE.find(&codec) {
        let mut parts = vec![
            parse_audio_codec(&codec[..m.start()]),
            codec[m.start()..m.end()].to_owned(),
        ];

        let left = parse_audio_codec(&codec[m.end()..]);
        if !left.is_empty() {
            parts.push(left);
        }

        parts.join(".")
    } else {
        parse_audio_codec(&codec)
    }
}

/// Helper function to parse audio codec
fn parse_audio_codec(codec: &str) -> String {
    if codec.is_empty() {
        return codec.to_owned();
    }

    let mut parts = Vec::new();
    let val = codec.replace(" ", ".");
    for p in val.split('.') {
        let p = p.trim();
        if p.is_empty() {
            continue;
        }

        let p = match p {
            "TRUEHD" => "TrueHD",
            "ATMOS" => "Atmos",
            "DTSHD" => "DTS-HD",
            _ => p,
        };
        parts.push(p.to_owned());
    }

    parts.join(".")
}

/// Normalizes HDR string to a standard format
pub(super) fn normalize_hdr(hdr: &str) -> String {
    let hdr = hdr.to_uppercase().replace("-", "");
    if hdr.contains("DOLBY") || hdr == "DOVI" {
        "DV".to_owned()
    } else {
        hdr
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_normalize_quality() {
        assert_eq!(normalize_quality("web-dl"), "WEB-DL");
        assert_eq!(normalize_quality("webdl"), "WEB-DL");
        assert_eq!(normalize_quality("web-rip"), "WEBRip");
        assert_eq!(normalize_quality("webrip"), "WEBRip");
        assert_eq!(normalize_quality("bluray"), "BluRay");
        assert_eq!(normalize_quality("blu-ray"), "BluRay");
        assert_eq!(normalize_quality("bdrip"), "BDRip");
        assert_eq!(normalize_quality("bd-rip"), "BDRip");
        assert_eq!(normalize_quality("brrip"), "BRRip");
        assert_eq!(normalize_quality("br-rip"), "BRRip");
        assert_eq!(normalize_quality("remux"), "Remux");
        assert_eq!(normalize_quality("unknown"), "unknown");
    }

    #[test]
    fn test_normalize_video_codec() {
        assert_eq!(normalize_video_codec("X264"), "H264");
        assert_eq!(normalize_video_codec("H.264"), "H264");
        assert_eq!(normalize_video_codec("AVC"), "H264");
        assert_eq!(normalize_video_codec("X265"), "H265");
        assert_eq!(normalize_video_codec("H.265"), "H265");
        assert_eq!(normalize_video_codec("HEVC"), "H265");
        assert_eq!(normalize_video_codec("AV1"), "AV1");
    }

    #[test]
    fn test_normalize_audio_codec() {
        assert_eq!(normalize_audio_codec("TRUEHD"), "TrueHD");
        assert_eq!(normalize_audio_codec("TRUEHD.7.1"), "TrueHD.7.1");
        assert_eq!(
            normalize_audio_codec("TRUEHD.7.1.ATMOS"),
            "TrueHD.7.1.Atmos"
        );
        assert_eq!(normalize_audio_codec("DTS-HD.MA.7.1"), "DTS-HD.MA.7.1");
        assert_eq!(normalize_audio_codec("DTSHD.MA.7.1"), "DTS-HD.MA.7.1");
    }

    #[test]
    fn test_normalize_hdr() {
        assert_eq!(normalize_hdr("DOLBY"), "DV");
        assert_eq!(normalize_hdr("DOLBY-VISION"), "DV");
        assert_eq!(normalize_hdr("DOVI"), "DV");
        assert_eq!(normalize_hdr("HDR10"), "HDR10");
        assert_eq!(normalize_hdr("HDR10+"), "HDR10+");
    }
}
